File: kafka-0.7.0/kafka-0.7.0-core/src/main/scala/kafka/log/LogManager.scala
/**
 * Licensed to the Apache Software Foundation (ASF) under one or more
 * contributor license agreements.  See the NOTICE file distributed with
 * this work for additional information regarding copyright ownership.
 * The ASF licenses this file to You under the Apache License, Version 2.0
 * (the "License"); you may not use this file except in compliance with
 * the License.  You may obtain a copy of the License at
 * 
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package kafka.log

import java.io._
import org.apache.log4j.Logger
import kafka.utils._
import scala.actors.Actor
import scala.collection._
import java.util.concurrent.CountDownLatch
import kafka.server.{KafkaConfig, KafkaZooKeeper}
import kafka.common.{InvalidTopicException, InvalidPartitionException}

/**
 * The guy who creates and hands out logs
 */
@threadsafe
private[kafka] class LogManager(val config: KafkaConfig,
                                private val scheduler: KafkaScheduler,
                                private val time: Time,
                                val logCleanupIntervalMs: Long,
                                val logCleanupDefaultAgeMs: Long,
                                needRecovery: Boolean) {
  
  val logDir: File = new File(config.logDir)
  private val numPartitions = config.numPartitions
  private val maxSize: Long = config.logFileSize  // 日志分片的最大值，如果超过，就可能回滚
  private val flushInterval = config.flushInterval
  private val topicPartitionsMap = config.topicPartitionsMap
  private val logger = Logger.getLogger(classOf[LogManager])
  private val logCreationLock = new Object
  private val random = new java.util.Random
  private var kafkaZookeeper: KafkaZooKeeper = null
  private var zkActor: Actor = null
  private val startupLatch: CountDownLatch = if (config.enableZookeeper) new CountDownLatch(1) else null
  private val logFlusherScheduler = new KafkaScheduler(1, "kafka-logflusher-", false)
  private val logFlushIntervalMap = config.flushIntervalMap
  private val logRetentionMSMap = getLogRetentionMSMap(config.logRetentionHoursMap)
  private val logRetentionSize = config.logRetentionSize
  println("--------------------LogManager-------init-------------logDir---------:"+ logDir)
  /* Initialize a log for each subdirectory of the main log directory */
  private val logs = new Pool[String, Pool[Int, Log]]()
  if(!logDir.exists()) { // 判断日志目录是否存在
    logger.info("No log directory found, creating '" + logDir.getAbsolutePath() + "'")
    logDir.mkdirs() // 不存在就创建
  }
  if(!logDir.isDirectory() || !logDir.canRead()){ // 如果不是目录，或者不可读
    throw new IllegalArgumentException(logDir.getAbsolutePath() + " is not a readable log directory.")
  }
  val subDirs = logDir.listFiles() // 获取所有的子目录列表文件
  if(subDirs != null) {
    for(dir <- subDirs) { // 开始遍历
      if(!dir.isDirectory()) {
        logger.warn("Skipping unexplainable file '" + dir.getAbsolutePath() + "'--should it be there?")
      } else { // 如果是文件目录
        logger.info("Loading log '" + dir.getName() + "'")
        println("--------------------LogManager-------init-------------dir---------:"+ dir)
        val log = new Log(dir, maxSize, flushInterval, needRecovery) // 创建Log对象
        val topicPartion = Utils.getTopicPartition(dir.getName) // 获取topic 和 partition
        logs.putIfNotExists(topicPartion._1, new Pool[Int, Log]())   // 放入缓冲中
        val parts = logs.get(topicPartion._1)
        parts.put(topicPartion._2, log) // 分区和文件对象建立对应关系
      }
    }
  }
  
  /* Schedule the cleanup task to delete old logs */
  if(scheduler != null) {
    logger.info("starting log cleaner every " + logCleanupIntervalMs + " ms")    
    scheduler.scheduleWithRate(cleanupLogs, 60 * 1000, logCleanupIntervalMs)
  }

  if(config.enableZookeeper) {
    kafkaZookeeper = new KafkaZooKeeper(config, this)
    kafkaZookeeper.startup
    zkActor = new Actor {
      def act() {
        loop {
          receive {
            case topic: String =>
              try {
                kafkaZookeeper.registerTopicInZk(topic)
              }
              catch {
                case e => logger.error(e) // log it and let it go
              }
            case StopActor =>
              logger.info("zkActor stopped")
              exit
          }
        }
      }
    }
    zkActor.start
  }

  case object StopActor

  private def getLogRetentionMSMap(logRetentionHourMap: Map[String, Int]) : Map[String, Long] = {
    var ret = new mutable.HashMap[String, Long]
    for ( (topic, hour) <- logRetentionHourMap )
      ret.put(topic, hour * 60 * 60 * 1000L)
    ret
  }

  /**
   *  Register this broker in ZK for the first time.
   */
  def startup() {
    if(config.enableZookeeper) {// 是否使用外部的zk
      logger.info("----------LogManager--------startup------------使用外部的zk服务-------------------")
      kafkaZookeeper.registerBrokerInZk()
      for (topic <- getAllTopics)
        kafkaZookeeper.registerTopicInZk(topic)
      startupLatch.countDown
    }
    logger.info("Starting log flusher every " + config.flushSchedulerThreadRate + " ms with the following overrides " + logFlushIntervalMap)
    logFlusherScheduler.scheduleWithRate(flushAllLogs, config.flushSchedulerThreadRate, config.flushSchedulerThreadRate)
  }

  private def awaitStartup() {
    if (config.enableZookeeper)
      startupLatch.await
  }

  def registerNewTopicInZK(topic: String) {
    if (config.enableZookeeper)
      zkActor ! topic 
  }

  /**
   * Create a log for the given topic and the given partition
   */
  private def createLog(topic: String, partition: Int): Log = {
    println("-------------LogManager--------------createLog-------------根据topic和分区--创建日志目录-----------")
    logCreationLock synchronized {
      val d = new File(logDir, topic + "-" + partition)  //  创建日志目录
      d.mkdirs()
      new Log(d, maxSize, flushInterval, false)
    }
  }
  

  def chooseRandomPartition(topic: String): Int = {
    random.nextInt(topicPartitionsMap.getOrElse(topic, numPartitions))
  }

  /**
   * Create the log if it does not exist, if it exists just return it
   */
  def getOrCreateLog(topic: String, partition: Int): Log = { // 需要根据topic和 partition分区，创建Log对象
    println("-------------LogManager----getOrCreateLog--------------------------")
    awaitStartup
    if (topic.length <= 0){
      throw new InvalidTopicException("topic name can't be empty")
    }
    if (partition < 0 || partition >= topicPartitionsMap.getOrElse(topic, numPartitions)) {
      logger.warn("Wrong partition " + partition + " valid partitions (0," + (topicPartitionsMap.getOrElse(topic, numPartitions) - 1) + ")")
      throw new InvalidPartitionException("wrong partition " + partition)
    }
    var hasNewTopic = false
    var parts: Pool[Int, Log] = logs.get(topic) // 根据topic从缓存池中，获取是否存在对应的分区的日志分片
    if (parts == null) { // 如果不存在，就new一个对象，存储到池子中
      val found = logs.putIfNotExists(topic, new Pool[Int, Log])
      if (found == null) {
        hasNewTopic = true // 说明是新的topic
      }
      parts = logs.get(topic) // 再次获取
    }
    var log = parts.get(partition) // 然后根据分区获取，log对象实例
    if(log == null) {  // 如果没有日志文件目录，就创建一个
      log = createLog(topic, partition)
      val found = parts.putIfNotExists(partition, log)
      if(found != null) {
        // there was already somebody there
        log.close()
        log = found
      } else{
        logger.info("Created log for '" + topic + "'-" + partition)
      }
    }

    if (hasNewTopic){ // 如果是新的topic，就注册到zk
      registerNewTopicInZK(topic)
    }
    log
  }
  
  /* Attemps to delete all provided segments from a log and returns how many it was able to */
  private def deleteSegments(log: Log, segments: Seq[LogSegment]): Int = {
    var total = 0
    for(segment <- segments) {
      logger.info("Deleting log segment " + segment.file.getName() + " from " + log.name)
      Utils.swallow(logger.warn, segment.messageSet.close())
      if(!segment.file.delete()) {
        logger.warn("Delete failed.")
      } else {
        total += 1
      }
    }
    total
  }

  /* Runs through the log removing segments older than a certain age */
  private def cleanupExpiredSegments(log: Log): Int = {
    println(Thread.currentThread().getName + "-------------LogManager-------------------cleanupExpiredSegments-----------start-------")
    val startMs = time.milliseconds
    val topic = Utils.getTopicPartition(log.dir.getName)._1
    val logCleanupThresholdMS = logRetentionMSMap.get(topic).getOrElse(this.logCleanupDefaultAgeMs)
    // 删除过期的文件，如果文件最后修改时间和当前时间做时间差，大于设置时间，就满足条件，删除
    val toBeDeleted = log.markDeletedWhile(startMs - _.file.lastModified > logCleanupThresholdMS)
    val total = deleteSegments(log, toBeDeleted) // 删除特定的日志分片
    println(Thread.currentThread().getName + "-------------LogManager-------------------cleanupExpiredSegments-----------end-------")
    total
  }

  /**
   *  Runs through the log removing segments until the size of the log
   *  is at least logRetentionSize bytes in size
   */
  private def cleanupSegmentsToMaintainSize(log: Log): Int = {
    if(logRetentionSize < 0 || log.size < logRetentionSize) return 0
    var diff = log.size - logRetentionSize
    def shouldDelete(segment: LogSegment) = {
      if(diff - segment.size >= 0) {
        diff -= segment.size
        true
      } else {
        false
      }
    }
    val toBeDeleted = log.markDeletedWhile( shouldDelete )
    val total = deleteSegments(log, toBeDeleted)
    total
  }

  /**
   * Delete any eligible logs. Return the number of segments deleted.
   */
  def cleanupLogs() {
    println(Thread.currentThread().getName + "-------------LogManager--------定时清理日志线程-----------cleanupLogs-----------start-------")
    logger.debug("Beginning log cleanup...")
    val iter = getLogIterator
    var total = 0
    val startMs = time.milliseconds
    while(iter.hasNext) {
      val log = iter.next
      logger.debug("Garbage collecting '" + log.name + "'")
      total += cleanupExpiredSegments(log) + cleanupSegmentsToMaintainSize(log)
    }
    logger.debug("Log cleanup completed. " + total + " files deleted in " + (time.milliseconds - startMs) / 1000 + " seconds")
    println(Thread.currentThread().getName +"-------------LogManager----------定时清理日志线程---------cleanupLogs-----------end-------")
  }
  
  /**
   * Close all the logs
   */
  def close() {
    logFlusherScheduler.shutdown()
    val iter = getLogIterator
    while(iter.hasNext)
      iter.next.close()
    if (config.enableZookeeper) {
      zkActor ! StopActor
      kafkaZookeeper.close
    }
  }
  
  private def getLogIterator(): Iterator[Log] = {
    new IteratorTemplate[Log] {
      val partsIter = logs.values.iterator
      var logIter: Iterator[Log] = null

      override def makeNext(): Log = {
        while (true) {
          if (logIter != null && logIter.hasNext)
            return logIter.next
          if (!partsIter.hasNext)
            return allDone
          logIter = partsIter.next.values.iterator
        }
        // should never reach here
        assert(false)
        return allDone
      }
    }
  }

  private def flushAllLogs() = {
    //println(Thread.currentThread().getName +"-------------LogManager---------flushAllLogs-----刷新日志-----------start-------")
    if (logger.isDebugEnabled)
      logger.debug("flushing the high watermark of all logs")

    for (log <- getLogIterator) {
      try{
        val timeSinceLastFlush = System.currentTimeMillis - log.getLastFlushedTime
        var logFlushInterval = config.defaultFlushIntervalMs
        if(logFlushIntervalMap.contains(log.getTopicName)) {
          logFlushInterval = logFlushIntervalMap(log.getTopicName)
        }
        if (logger.isDebugEnabled){
          logger.debug(log.getTopicName + " flush interval  " + logFlushInterval + " last flushed " + log.getLastFlushedTime + " timesincelastFlush: " + timeSinceLastFlush)
        }
        if(timeSinceLastFlush >= logFlushInterval) {
          log.flush
        }
      } catch {
        case e =>
          logger.error("Error flushing topic " + log.getTopicName, e)
          e match {
            case _: IOException =>
              logger.fatal("Halting due to unrecoverable I/O error while flushing logs: " + e.getMessage, e)
              Runtime.getRuntime.halt(1)
            case _ =>
          }
      }
    }
   // println(Thread.currentThread().getName +"-------------LogManager---------flushAllLogs-----刷新日志-----------end-------")
  }


  def getAllTopics(): Iterator[String] = logs.keys.iterator
  def getTopicPartitionsMap() = topicPartitionsMap
}

File: kafka-0.7.0/kafka-0.7.0-core-java/src/main/java/kafka/log/LogManager.java
package kafka.log;

import clover.org.apache.commons.lang.StringUtils;
import kafka.common.InvalidPartitionException;
import kafka.common.InvalidTopicException;
import kafka.server.KafkaConfig;
import kafka.server.KafkaZooKeeper;
import kafka.utils.*;
import org.apache.log4j.Logger;
import scala.Tuple2;

import java.io.File;
import java.io.IOException;
import java.util.*;
import java.util.concurrent.ConcurrentHashMap;
import java.util.concurrent.CountDownLatch;

public class LogManager {
    
    private Logger logger = Logger.getLogger(LogManager.class);
    private Object logCreationLock = new Object();
    private Random random = new java.util.Random();
    private KafkaZooKeeper kafkaZookeeper = null;
    private Thread zkActorThread = null;
    private KafkaScheduler logFlusherScheduler = new KafkaScheduler(1, "kafka-logflusher-", false);
    
    private Map<String, Integer> logFlushIntervalMap = null;
    private Map<String, Long> logRetentionMSMap = null;
    private int logRetentionSize;
    private CountDownLatch startupLatch = null;
    private File logDir = null;
    private int numPartitions;
    private Long maxSize = null;
    private int flushInterval;
    private Map<String, Integer> topicPartitionsMap;
    
    public KafkaConfig config;
    public KafkaScheduler scheduler;
    public Time time;
    public Long logCleanupIntervalMs;
    public Long logCleanupDefaultAgeMs;
    public Boolean needRecovery;
    
    private Pool logs = new Pool<String, Pool<Integer, Log>>();
    
    private volatile String topic = "";
    private volatile String state = "run";
    
    public LogManager(KafkaConfig config,
                      KafkaScheduler scheduler,
                      Time time,
                      Long logCleanupIntervalMs,
                      Long logCleanupDefaultAgeMs,
                      Boolean needRecovery) {
        this.config = config;
        this.scheduler = scheduler;
        this.time = time;
        this.logCleanupIntervalMs = logCleanupIntervalMs;
        this.logCleanupDefaultAgeMs = logCleanupDefaultAgeMs;
        this.needRecovery = needRecovery;
        
        
        this.logFlushIntervalMap = config.flushIntervalMap;
        this.logRetentionMSMap = getLogRetentionMSMap(config.logRetentionHoursMap);
        this.logRetentionSize = config.logRetentionSize;
        if (config.enableZookeeper) {
            this.startupLatch = new CountDownLatch(1);
        }
        this.logDir = new File(config.logDir);
        this.numPartitions = config.numPartitions;
        this.maxSize = Long.parseLong(config.logFileSize + "");
        this.flushInterval = config.flushInterval;
        this.topicPartitionsMap = config.topicPartitionsMap;
        
        // 初始化
        init();
    }
    
    
    void init() {
        if (!logDir.exists()) {
            logger.info("No log directory found, creating '" + logDir.getAbsolutePath() + "'");
            logDir.mkdirs();
        }
        if (!logDir.isDirectory() || !logDir.canRead())
            throw new IllegalArgumentException(logDir.getAbsolutePath() + " is not a readable log directory.");
        File subDirs[] = logDir.listFiles();
        if (subDirs != null) {
            for (File dir : subDirs) {
                if (!dir.isDirectory()) {
                    logger.warn("Skipping unexplainable file '" + dir.getAbsolutePath() + "'--should it be there?");
                } else {
                    logger.info("Loading log '" + dir.getName() + "'");
                    Log log = new Log(dir, maxSize, flushInterval, needRecovery);
                    Tuple2<String, Integer> topicPartion = Utils.getTopicPartition(dir.getName());
                    logs.putIfNotExists(topicPartion._1, new Pool<Integer, Log>());
                    Pool<Integer, Log> parts = (Pool<Integer, Log>) logs.get(topicPartion._1);
                    parts.put(topicPartion._2, log);
                }
            }
        }
        
        /* Schedule the cleanup task to delete old logs */
        if (scheduler != null) {
            logger.info("定时清除任务以删除旧日志 starting log cleaner every " + logCleanupIntervalMs + " ms");
            UnitFunction function = new UnitFunction() {
                @Override
                public void call() {
                    logger.info("定时清理日志线程：" + Thread.currentThread().getName());
                    cleanupLogs();
                }
                
                @Override
                public Object call(Object v) {
                    return null;
                }
            };
            scheduler.scheduleWithRate(function, 60 * 1000L, logCleanupIntervalMs);
        }
        
        if (config.enableZookeeper) {
            kafkaZookeeper = new KafkaZooKeeper(config, this);
            kafkaZookeeper.startup();

//            zkActorThread = new Thread() {
//                @Override
//                public void run() {
//                    while (true) {
//                        // 监听topic信息
//                        if (!StringUtils.isEmpty(topic)) {
//                            kafkaZookeeper.registerTopicInZk(topic);
//                        }
//
//                        if (state.equals("stop")) {
//                            Thread.currentThread().interrupt();
//                            logger.info("zkActor stopped");
//                            System.exit(0);
//                        }
//                    }
//                }
//            };
//            zkActorThread.start();
        }
        
    }
    
    
    private Map<String, Long> getLogRetentionMSMap(Map<String, Integer> logRetentionHourMap) {
        Map<String, Long> ret = new HashMap<String, Long>();
        for (String topic : logRetentionHourMap.keySet()) {
            Integer hour = logRetentionHourMap.get(topic);
            ret.put(topic, hour * 60 * 60 * 1000L);
        }
        return ret;
    }
    
    void cleanupLogs() {
        logger.debug("Beginning log cleanup...");
        Iterator<Log> iter = getLogIterator();
        int total = 0;
        Long startMs = time.milliseconds();
        while (iter.hasNext()) {
            Log log = iter.next();
            logger.debug("Garbage collecting '" + log.name + "'");
            total += cleanupExpiredSegments(log) + cleanupSegmentsToMaintainSize(log);
        }
        logger.debug("Log cleanup completed. " + total + " files deleted in " + (time.milliseconds() - startMs) / 1000 + " seconds");
    }
    
    private Integer cleanupSegmentsToMaintainSize(Log log) {
        if (logRetentionSize < 0 || log.size() < logRetentionSize) return 0;
        long diff = log.size() - logRetentionSize;
        
        
        List<LogSegment> toBeDeleted = log.markDeletedWhileV2(diff);
        int total = deleteSegments(log, toBeDeleted);
        return total;
    }
    
    
    private Integer cleanupExpiredSegments(Log log) {
        Long startMs = time.milliseconds();
        String topic = Utils.getTopicPartition(log.dir.getName())._1;
        Long logCleanupThresholdMS = logRetentionMSMap.get(topic);
        if (logCleanupThresholdMS == null) {
            logCleanupThresholdMS = this.logCleanupDefaultAgeMs;
        }
        // startMs - _.file.lastModified > logCleanupThresholdMS
        List<LogSegment> toBeDeleted = log.markDeletedWhile(startMs, logCleanupThresholdMS);
        int total = deleteSegments(log, toBeDeleted);
        return total;
    }
    
    private int deleteSegments(Log log, List<LogSegment> segments) {
        int total = 0;
        for (LogSegment segment : segments) {
            logger.info("Deleting log segment " + segment.file.getName() + " from " + log.name);
            
            try {
                segment.messageSet.close();
            } catch (IOException e) {
                e.printStackTrace();
                logger.warn(e.getMessage(), e);
            }
            
            if (!segment.file.delete()) {
                logger.warn("Delete failed.");
            } else {
                total += 1;
            }
        }
        return total;
    }
    
    
    private Iterator<Log> getLogIterator() {
        return new IteratorTemplate<Log>() {
            Iterator<Pool> partsIter = logs.values.iterator();
            Iterator<Log> logIter = null;
            
            @Override
            protected Log makeNext() throws Throwable {
                while (true) {
                    if (logIter != null && logIter.hasNext())
                        return logIter.next();
                    if (!partsIter.hasNext())
                        return allDone();
                    Pool pool = partsIter.next();
                    logIter = pool.values.iterator();
                }
                // should never reach here
            }
        };
    }
    
    
    void test(Boolean assertion) {
        if (!assertion)
            throw new java.lang.AssertionError("assertion failed");
    }
    
    
    public Map getTopicPartitionsMap() {
        return topicPartitionsMap;
    }
    
    private void awaitStartup() {
        if (config.enableZookeeper)
            try {
                startupLatch.await();
            } catch (InterruptedException e) {
                e.printStackTrace();
            }
    }
    
    public Log getOrCreateLog(String topic, int partition) {
        awaitStartup();
        if (topic.length() <= 0) {
            throw new InvalidTopicException("topic name can't be empty");
        }
        Integer p = topicPartitionsMap.get(topic);
        if (p == null) {
            p = numPartitions;
        }
        if (partition < 0 || partition > p) {
            logger.warn("Wrong partition " + partition + " valid partitions (0," + (p - 1) + ")");
            throw new InvalidPartitionException("wrong partition " + partition);
        }
        boolean hasNewTopic = false;
        Pool<Integer, Log> parts = (Pool<Integer, Log>) logs.get(topic);
        if (parts == null) {
            Pool<Integer, Log> found = (Pool<Integer, Log>) logs.putIfNotExists(topic, new Pool<Integer, Log>());
            if (found == null)
                hasNewTopic = true;
            parts = (Pool<Integer, Log>) logs.get(topic);
        }
        Log log = parts.get(partition);
        if (log == null) {
            log = createLog(topic, partition);
            Log found = parts.putIfNotExists(partition, log);
            if (found != null) {
                // there was already somebody there
                log.close();
                log = found;
            } else
                logger.info("Created log for '" + topic + "'-" + partition);
        }
        
        if (hasNewTopic) {
            registerNewTopicInZK(topic);
        }
        return log;
    }
    
    public void registerNewTopicInZK(String topic) {
        if (config.enableZookeeper)
            this.topic = topic;
    }
    
    private Log createLog(String topic, int partition) {
        synchronized (logCreationLock) {
            File d = new File(logDir, topic + "-" + partition);
            d.mkdirs();
            return new Log(d, maxSize, flushInterval, false);
        }
    }
    
    public int chooseRandomPartition(String topic) {
        Integer partition = topicPartitionsMap.get(topic);
        if (partition == null) {
            partition = numPartitions;
        }
        return random.nextInt(partition);
    }
    
    
    Iterator<String> getAllTopics() {
        return logs.keys.iterator();
    }
    
    public void startup() {
        if (config.enableZookeeper) {
            kafkaZookeeper.registerBrokerInZk();
            Iterator<String> iterator = getAllTopics();
            while (iterator.hasNext()) {
                String topic = iterator.next();
                kafkaZookeeper.registerTopicInZk(topic);
            }
            startupLatch.countDown();
        }
        logger.info("Starting log flusher every " + config.flushSchedulerThreadRate + " ms with the following overrides " + logFlushIntervalMap);
        UnitFunction function = new UnitFunction() {
            @Override
            public void call() {
                flushAllLogs();
            }
            
            @Override
            public Object call(Object v) {
                return null;
            }
        };
        
        logFlusherScheduler.scheduleWithRate(function, Long.parseLong(config.flushSchedulerThreadRate + ""), Long.parseLong(config.flushSchedulerThreadRate + ""));
    }
    
    private void flushAllLogs() {
        if (logger.isDebugEnabled())
            logger.debug("flushing the high watermark of all logs");
        Iterator<Log> iterator = getLogIterator();
        while (iterator.hasNext()) {
            Log log = iterator.next();
            try {
                int timeSinceLastFlush = (int) (System.currentTimeMillis() - log.getLastFlushedTime());
                int logFlushInterval = config.defaultFlushIntervalMs;
                if (logFlushIntervalMap.containsKey(log.getTopicName()))
                    logFlushInterval = logFlushIntervalMap.get(log.getTopicName());
                if (logger.isDebugEnabled())
                    logger.debug(log.getTopicName() + " flush interval  " + logFlushInterval +
                            " last flushed " + log.getLastFlushedTime() + " timesincelastFlush: " + timeSinceLastFlush);
                if (timeSinceLastFlush >= logFlushInterval)
                    log.flush();
            } catch (Exception e) {
                logger.error("Error flushing topic " + log.getTopicName(), e);
                logger.fatal("Halting due to unrecoverable I/O error while flushing logs: " + e.getMessage(), e);
                Runtime.getRuntime().halt(1);
            }
        }
    }
    
    
    public void close() {
        logFlusherScheduler.shutdown();
        Iterator<Log> iter = getLogIterator();
        while (iter.hasNext())
            iter.next().close();
        if (config.enableZookeeper) {
            state = "stop";
            kafkaZookeeper.close();
        }
    }
    
}
